Tests/ControllersTest/MessageLabelSpec.swift

```diff
@@ -0,0 +1,258 @@
+                it("updates the lineBreakMode of attributedText") {
+
+                }
+            }
+            context("when the textAlignment property is set") {
+                it("updates the textAlignment of attributedText") {
+
+                }
+            }
+        }
+
+        describe("updating the attributes of detected text") {
+            context("addressAttributes is set to a new value") {
+                it("updates ranges for detected addresses") {
+
+                }
+            }
+            context("phoneNumberAttributes is set to a new value") {
+                it("updates ranges for detected phone numbers") {
+
+                }
+            }
+            context("dateAttributes is set to a new value") {
+                it("updates ranges for detected dates") {
+
+                }
+            }
+            context("urlAttributes is set to a new value") {
+                it("updates ranges for detected urls") {
+
+                }
+            }
+        }
+
+        describe("text dectection of multiple DetectorTypes") {
+            context("contains address, date, url, and phone number") {
+                it("applies addressAttributes to detected addresses") {
+
+                }
+                it("applies phoneNumberAttributes to detected phone numbers") {
+
+                }
+                it("applies urlAttributes to detected urls") {
+
+                }
+                it("applies dateAttributes to detected dates") {
+
+                }
+                it("does not apply attributes to non detected ranges") {
+                    
+                }
+            }
+        }
+
+        describe("touch handling for detected text") {
+            context("message text contains detected address") {
+                context("touch occurs on address") {
+
+                }
+                context("touch does not occur on address") {
+
+                }
+            }
+            context("message text contains detected phone number") {
+                context("touch occurs on phone number") {
+
+                }
+                context("touch does not occur on phone number") {
+
+                }
+            }
+            context("message text contains detected url") {
+                context("touch occurs on url") {
+
+                }
+                context("touch does not occur on url") {
+
+                }
+            }
+            context("message text contains detected date") {
+                context("touch occurs on date") {
+
+                }
+                context("touch does not occur on date") {
+
+                }
+            }
+        }
+    }
+}
+
+// MARK: - Helpers
+
+fileprivate extension MessageLabel {
+
+    var textAttributes: [NSAttributedStringKey: Any]? {
+        let length = attributedText!.length
+        var range = NSRange(location: 0, length: length)
+        return attributedText?.attributes(at: 0, effectiveRange: &range)
```

Comment:
You treat `attributedText` here as an Optional, but the length value above is calculated with the value being force-unwrapped. This means either the function will crash if it's nil, or doesn't ever need to return an optional if you're force-unwrapping it!

Reply:
Harsh critic 😂 😂

Reply:
❤️